### PR TITLE
combine `ifelse` and `select_value`; improve `ifelse` t-func

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

**Comment by Keno** [comment]
Should address the same benchmarks that #26969 was supposed to.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27068#issuecomment-388221374) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/64e8ca95cd19781e4a51b61a2af1b82aa8a38e76/431a06c_vs_f1b2328/report.md). cc @ararslan

**Comment by JeffBezanson** [comment]
Do we think this actually introduced allocations in the union benchmarks, or is something up with the benchmarks or the comparison data?

**Comment by Keno** [comment]
I was about to check.

**Comment by Keno** [comment]
Looks like something is broken with select_value codegen. It boxes the value in order to do the select:
```
  %26 = bitcast %jl_value_t addrspace(11)* %23 to float addrspace(11)*
  %27 = load float, float addrspace(11)* %26, align 4
  %28 = call %jl_value_t addrspace(10)* @jl_box_float32(float %27)
  %29 = select i1 %25, %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 140374922305472 to %jl_value_t*) to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)* %28
; Function +; {
; Location: float.jl:390
  %30 = bitcast %jl_value_t addrspace(10)* %29 to float addrspace(10)*
  %31 = load float, float addrspace(10)* %30, align 4
  %32 = fadd fast float %s.225, %31
```
I suspect you need to use the inferred return type in codegen to force an implicit convert of the arguments

**Comment by Keno** [comment]
I have a fix. Let me push it here.

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

File: src/processor.cpp [hunk 1]
```diff
@@ -783,8 +783,10 @@ static inline SysimgMatch match_sysimg_targets(S &&sysimg, T &&target, F &&max_v
         match.best_idx = i;
         feature_size = new_feature_size;
     }
-    if (match.best_idx == (uint32_t)-1)
-        jl_error("Unable to find compatible target in system image.");
+    if (match.best_idx == (uint32_t)-1) {
+        match.best_idx = 0;
+        //jl_error("Unable to find compatible target in system image.");
+    }
```
Comment (JeffBezanson) [inline src/processor.cpp]: ?

Comment (Keno) [reply src/processor.cpp]: whoops.

**Comment by Keno** [comment]
Fixed. That's what I get for coding before breakfast.

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

File: base/compiler/tfuncs.jl [hunk 2]
```diff
@@ -200,8 +200,9 @@ add_tfunc(ifelse, 3, 3,
             else
                 return Bottom
             end
-        end
-        if !isa(cnd, Conditional) && !(Bool ⊑ cnd)
+        elseif isa(cnd, Conditional)
+            # handled in abstract_call
```
Comment (vtjnash) [inline base/compiler/tfuncs.jl]: it's not

File: base/compiler/tfuncs.jl [hunk 3]
```diff
@@ -200,8 +200,11 @@ add_tfunc(Core.Intrinsics.select_value, 3, 3,
             else
                 return Bottom
             end
+        elseif isa(cnd, Conditional)
+            # handled in abstract_call
```
[the remark this answers is not in the document]
Comment (JeffBezanson) [reply base/compiler/tfuncs.jl]: Is there anything further that can be done with `Conditional` here? Anyway I think it's helpful to have this comment so one knows where to look for the full behavior.

[the remark this answers is not in the document]
Comment (Keno) [reply base/compiler/tfuncs.jl]: you can check whether the conditional is equivalent to a Const (i.e. that special case we added in).

[the remark this answers is not in the document]
Comment (JeffBezanson) [reply base/compiler/tfuncs.jl]: Ok. I say we do that later unless it's needed to meet our immediate goals.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27068#issuecomment-388390746) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/59648be7d76140c552d4a504d0bcfb4adb1b867d/92024e5_vs_f90c599/report.md). cc @ararslan

**Comment by Keno** [comment]
Performance looks good.

**Pull request merged**